receiver/pluginreceiver/plugin.go

```diff
@@ -0,0 +1,211 @@
+	}
+
+	var plugin Plugin
+	if err := yaml.Unmarshal(bytes, &plugin); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal plugin from yaml: %w", err)
+	}
+
+	return &plugin, nil
+}
+
+// RenderComponents renders the plugin's template as a component map
+func (p *Plugin) RenderComponents(values map[string]interface{}) (*ComponentMap, error) {
+	template, err := template.New(p.Title).Parse(p.Template)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create plugin template: %w", err)
+	}
+
+	templateValues := p.ApplyDefaults(values)
+
+	var writer bytes.Buffer
+	if err := template.Execute(&writer, templateValues); err != nil {
+		return nil, fmt.Errorf("failed to execute template: %w", err)
+	}
+
+	return unmarshalComponentMap(writer.Bytes())
+}
+
+// ApplyDefaults returns a copy of the values map with parameter defaults applied.
+// If a value is already present in the map, it supercedes the default.
+func (p *Plugin) ApplyDefaults(values map[string]interface{}) map[string]interface{} {
+	result := make(map[string]interface{})
+
+	for _, parameter := range p.Parameters {
+		if parameter.Default == nil {
+			continue
+		}
+
+		result[parameter.Name] = parameter.Default
+	}
+
+	for key, value := range values {
+		result[key] = value
+	}
+
+	return result
+}
+
+// CheckParameters checks the supplied values against the defined parameters of the plugin
+func (p *Plugin) CheckParameters(values map[string]interface{}) error {
+	if err := p.checkDefined(values); err != nil {
+		return fmt.Errorf("definition failure: %w", err)
+	}
+
+	if err := p.checkRequired(values); err != nil {
+		return fmt.Errorf("required failure: %w", err)
+	}
+
+	if err := p.checkType(values); err != nil {
+		return fmt.Errorf("type failure: %w", err)
+	}
+
+	if err := p.checkSupported(values); err != nil {
+		return fmt.Errorf("supported value failure: %w", err)
+	}
+
+	return nil
+}
+
+// checkDefined checks if any of the supplied values are defined by the plugin
+func (p *Plugin) checkDefined(values map[string]interface{}) error {
+	parameterMap := make(map[string]Parameter)
+	for _, parameter := range p.Parameters {
+		parameterMap[parameter.Name] = parameter
+	}
+
+	for key := range values {
+		if _, ok := parameterMap[key]; !ok {
+			return fmt.Errorf("parameter %s is not defined in plugin", key)
+		}
+	}
+
+	return nil
+}
+
+// checkRequired checks if required values are defined
+func (p *Plugin) checkRequired(values map[string]interface{}) error {
+	for _, parameter := range p.Parameters {
+		_, ok := values[parameter.Name]
+		if parameter.Required && !ok {
+			return fmt.Errorf("parameter %s is missing but required in plugin", parameter.Name)
+		}
+	}
+
+	return nil
+}
+
+// checkType checks if the values match their parameter type
+func (p *Plugin) checkType(values map[string]interface{}) error {
+	for _, parameter := range p.Parameters {
+		value, ok := values[parameter.Name]
+		if !ok {
+			continue
+		}
+
+		switch parameter.Type {
```

Comment:
Will go limit you to only specifying parameter types that were defined in your const list? I don't remember but since it's just a string alias I feel like you can probably put any valid string as a parameter.Type. If that's the case we should probably have a default clause here.